webview-ui/src/components/chat/chat-view/components/layout/ActionButtons.tsx

```diff
@@ -17,26 +22,107 @@ interface ActionButtonsProps {
 /**
  * Action buttons area including scroll-to-bottom and approve/reject buttons
  */
-export const ActionButtons: React.FC<ActionButtonsProps> = ({ chatState, messageHandlers, isStreaming, scrollBehavior }) => {
-	const { primaryButtonText, secondaryButtonText, enableButtons, didClickCancel, inputValue, selectedImages, selectedFiles } =
-		chatState
+export const ActionButtons: React.FC<ActionButtonsProps> = ({
+	task,
+	messages,
+	chatState,
+	mode,
+	messageHandlers,
+	scrollBehavior,
+}) => {
+	const { didClickCancel, inputValue, selectedImages, selectedFiles, setSendingDisabled, setDidClickCancel } = chatState
+
+	const [isStreaming, setIsStreaming] = useState(task?.partial === true)
+
+	const [primaryButtonText, setPrimaryButtonText] = useState<string | undefined>(undefined)
+	const [secondaryButtonText, setSecondaryButtonText] = useState<string | undefined>(undefined)
+
+	const [enableButtons, setEnableButtons] = useState<boolean>(false)
+
+	const [lastMessage, secondLastMessage] = useMemo(() => {
+		return [messages.at(-1), messages.at(-2)]
+	}, [messages])
+
+	// Memoize the button configuration to avoid recalculation
+	const buttonConfig = useMemo(() => {
+		// Special case: api_req_started after command_output needs special handling
+		if (lastMessage?.type === "say" && lastMessage.say === "api_req_started" && secondLastMessage?.ask === "command_output") {
+			return BUTTON_CONFIGS.api_req_active
+		}
+
+		return getButtonConfig(lastMessage, mode)
+	}, [lastMessage, secondLastMessage, mode])
+
+	// Apply button configuration with a single batched update
+	useEffect(() => {
+		// Batch all state updates together to prevent intermediate renders
+		setEnableButtons(buttonConfig.enableButtons)
+		setSendingDisabled(buttonConfig.sendingDisabled)
+		setPrimaryButtonText(buttonConfig.primaryText)
+		setSecondaryButtonText(buttonConfig.secondaryText)
+
+		// Handle special state changes for resume tasks
+		if (task?.ask === "resume_task" || task?.ask === "resume_completed_task") {
+			setDidClickCancel(false)
+		} else {
+			setIsStreaming(task?.partial === true)
+		}
+	}, [buttonConfig, task?.ask, task?.partial, setSendingDisabled, setDidClickCancel])
+
+	// Clear input when transitioning from command_output to api_req
+	// This happens when user provides feedback during command execution
+	useEffect(() => {
+		if (lastMessage?.type === "say" && lastMessage.say === "api_req_started" && secondLastMessage?.ask === "command_output") {
+			if (chatState.inputValue) {
+				chatState.setInputValue("")
+			}
+			if (chatState.selectedImages.length) {
+				chatState.setSelectedImages([])
+			}
+			if (chatState.selectedFiles.length) {
+				chatState.setSelectedFiles([])
+			}
```

Comment:
Can you also clear the cancel 
<img width="364" height="282" alt="image" src="https://github.com/user-attachments/assets/edb35901-d365-482e-acf9-764ccf3f31b0" />

Reply:
So when you start a new request this should be cleared. Because once you set it to true when do we cancel it?

Reply:
thanks for clarifying! I think this shows the issue with how we are managing states in different places that show be unified into a centralized place! I have removed the `setDidClickCancel`  altogether.